packages/react-reconciler/src/ReactFiberWorkLoop.js

```diff
@@ -644,26 +644,28 @@ function performConcurrentWorkOnRoot(root, didTimeout) {
   // event time. The next update will compute a new event time.
   currentEventTime = NoWork;
 
-  // Check if the render expired. If so, restart at the current time so that we
-  // can finish all the expired work in a single batch. However, we should only
-  // do this if we're starting a new tree. If we're in the middle of an existing
-  // tree, we'll continue working on that (without yielding) so that the work
-  // doesn't get dropped. If there's another expired level after that, we'll hit
-  // this path again, at which point we can batch all the subsequent levels
-  // together.
-  if (didTimeout && workInProgress === null) {
-    // Mark the current time as expired to synchronously render all expired work
-    // in a single batch.
-    const currentTime = requestCurrentTimeForUpdate();
-    markRootExpiredAtTime(root, currentTime);
+  // Check if the render expired.
+  const expirationTime = getNextRootExpirationTimeToWorkOn(root);
+  if (didTimeout) {
+    if (workInProgressRoot === root && renderExpirationTime !== NoWork) {
+      // We're already in the middle of working on this tree. Expire the tree at
+      // the already-rendering time so we don't throw out the partial work. If
+      // there's another expired level after that, we'll hit the `else` branch
+      // of this condition, which will batch together the remaining levels.
+      markRootExpiredAtTime(root, renderExpirationTime);
```

Comment:
It seems a bit like this make the expired value lie. If we added more places where we check the current time, we would update this to something further out. It's also not clear to me whether when we're actually going to set its to its real value and stop temporarily lying.

Instead of doing this here. Can we just let it set it to the lower expired value like we did before? Then instead check in performSyncWorkOnRoot if it has expired work but the expired work is lower priority then the currently rendering work, first do a pass finishing the currently rendering work and then do a second pass at the expired priority and then possibly a third pass if it errors in the second pass?

https://github.com/facebook/react/blob/master/packages/react-reconciler/src/ReactFiberWorkLoop.js#L992